### PR TITLE
Support for Streams from query objects

### PR DESCRIPTION
**Pull request by Buggaboo**
@greenrobot-team Please review.

### CODE REVIEW

**Comment by greenrobot-team** [comment]
@Buggaboo Thanks. Chiming in here: similar to the other bindings (e.g. Java https://docs.objectbox.io/data-observers-and-rx#data-observers-basics) observers are meant to notify that there was a change to a box (not what has changed) or more importantly return updated query results if there was a change to a box.

The internal APIs only provide ways to listen that there was a change to a single box or all boxes, not which changes. The binding should then add wrapper code so the user can subscribe to e.g. the latest query results. Again, see the Java binding for examples.
https://docs.objectbox.io/data-observers-and-rx#data-observers-basics

Hope this helps.

**Comment by Buggaboo** [comment]
> @Buggaboo Thanks. Chiming in here: similar to the other bindings (e.g. Java https://docs.objectbox.io/data-observers-and-rx#data-observers-basics) observers are meant to notify that there was a change to a box (not what has changed) or more importantly return updated query results if there was a change to a box.
> 
> The internal APIs only provide ways to listen that there was a change to a single box or all boxes, not which changes. The binding should then add wrapper code so the user can subscribe to e.g. the latest query results (if there was a change, re-run the query and return the new results). Again, see the Java binding for examples.
> https://docs.objectbox.io/data-observers-and-rx#data-observers-basics
> 
> Hope this helps.

Thanks, I'll look into it again, and try to follow the existing syntax.

File: lib/src/observable.dart [hunk 1]
```diff
@@ -0,0 +1,76 @@
...
+// dart callback signature
+typedef Any = void Function(Pointer<Void>, Pointer<Uint32>, int);
+
+class Observable {
+
+  static final anyObserver = <int, Pointer<Void>>{};
+  static final any = <int, Any>{}; // radix? > tree?
+
+  // sync:true -> ObjectBoxException: 10001 TX is not active anymore: #101
+  static final controller = StreamController<int>.broadcast();
```
[the remark this answers is not in the document]
Comment (Buggaboo) [reply lib/src/observable.dart:20]: Yeah, that requires some attention

File: lib/src/query/query.dart [hunk 2]
```diff
@@ -533,11 +533,12 @@ class ConditionGroupAll extends ConditionGroup {
 
 class Query<T> {
   Pointer<Void> _cQuery;
-  Store _store;
+  Store store;
```
[the remark this answers is not in the document]
Comment (Buggaboo) [reply lib/src/query/query.dart:536]: I need it for the extension

File: README.md [hunk 3]
```diff
@@ -148,6 +148,37 @@ scoreQuery.close();
 query.close();
 ```
 
+### Streams
+
+Streams can be created from queries.
+The streams can be extended with [rxdart](https://github.com/ReactiveX/rxdart);
```
[the remark this answers is not in the document]
Comment (Buggaboo) [reply README.md:154]: It seems to be the most popular one.

**Comment by greenrobot-team** [comment]
@Buggaboo A general remark: please look at the Dart Analysis issues and resolve them if possible (e.g. compare to changes made to your last PR).

File: lib/src/observable.dart [hunk 4]
```diff
@@ -0,0 +1,76 @@
...
+    for(var i=0; i<mutated_count; i++) {
+      // call schema's callback
+      if (any.containsKey(mutated_ids[i])) {
+        any[mutated_ids[i]](user_data, mutated_ids, mutated_count);
+      }
+    }
+  }
+
+  static void subscribe(Store store) {
+    final callback = Pointer.fromFunction<obx_observer_t<Void, Uint32>>(_anyCallback);
```
Comment (greenrobot-team) [inline lib/src/observable.dart:32]: This does not appear right (Dart error). Shouldn't the second type be `Pointer<Uint32>`? Then `_anyCallback` needs updating as well.
Anyhow, I would recommend avoiding type args if not really necessary when defining the functions in `signatures.dart` (`typedef obx_observer_t` etc.). They make things needlessly complicated.

File: lib/src/bindings/signatures.dart [hunk 5]
```diff
@@ -159,6 +159,14 @@ typedef obx_query_visit_native_t = Int32 Function(Pointer<Void> query,
 typedef obx_query_visit_dart_t = int Function(Pointer<Void> query,
     Pointer<NativeFunction<obx_data_visitor_native_t>> visitor, Pointer<Void> user_data, int offset, int limit);
 
+// observers
+
+typedef obx_observer_t<U extends NativeType, T extends NativeType> = U Function(Pointer<Void> user_data, Pointer<Uint32> entity_id, T type_ids_count);
+typedef obx_observer_single_type_t<U extends NativeType> = U Function(Pointer<Void> user_data);
+typedef obx_observe_t = Pointer<Void> Function(Pointer<Void> store, Pointer<NativeFunction<obx_observer_t<Void, Uint32>>> callback, Pointer<Void> user_data);
+typedef obx_observe_single_type_t<T> = Pointer<Void> Function(Pointer<Void> store, T entity_id, Pointer<NativeFunction<obx_observer_single_type_t<Void>>> callback, Pointer<Void> user_data);
+typedef obx_observer_close_t<U extends NativeType> = U Function(Pointer<Void> observer);
```
Comment (greenrobot-team) [inline lib/src/bindings/signatures.dart:168]: As said most of these use AFAIC tell unnecessary type args. Is there any reason for this complication?
E.g.
```
typedef obx_observer_t = Void Function(Pointer<Void> user_data, Pointer<Uint32> entity_id, Pointer<Uint32> type_ids_count);
```
should be enough?

Comment (Buggaboo) [reply lib/src/bindings/signatures.dart:168]: sorry, real life got in the way, I made the changes!

Comment (greenrobot-team) [reply lib/src/bindings/signatures.dart]: Thanks, I'll have a look once I can make some time.

File: lib/src/observable.dart [hunk 6]
```diff
@@ -0,0 +1,76 @@
...
+  void unsubscribe () { Observable.unsubscribe(this); }
+}
+
+extension Streamable<T> on Query<T> {
+  void _setup() {
+    if (!Observable.anyObserver.containsKey(store.ptr)) {
+      store.subscribe();
+    }
+
+    // Assume consensus on entityId over all available Stores
```
Comment (greenrobot-team) [inline lib/src/observable.dart:57]: Are multiple Stores supported by the Dart library? If yes, the entity ID is not globally unique. So this won't work?

File: lib/src/observable.dart (hunk 1 above)
[the remark this answers is not in the document]
Comment (greenrobot-team) [reply lib/src/observable.dart:20]: What does the reference of #101 mean here?

File: test/stream_test.dart [hunk 7]
```diff
@@ -0,0 +1,69 @@
...
+      result.add(str);
+    });
+
+    box.put(TestEntity(tString: 'Hello world'));
+    box.putMany(<TestEntity>[ TestEntity(tString: 'Goodbye'),
+      TestEntity(tString: 'for now') ]);
+
+    await Future.delayed(Duration(seconds: 0)); // ffi explodes without
+    expect(result,
+        ['for now, Goodbye, Hello world', 'for now, Goodbye, Hello world']);
```
Comment (greenrobot-team) [inline test/stream_test.dart:37]: Shouldn't this be `['Hello world', 'for now, Goodbye, Hello world']? E.g. first put is one event, second putMany the other. Or am I missing something here?

File: test/stream_test.dart [hunk 8]
```diff
@@ -0,0 +1,69 @@
...
+    });
+
+    box.put(TestEntity(tString: 'Hello world'));
+
+    // idem, see above
+    box.putMany(<TestEntity>[ TestEntity(tString: 'Goodbye'),
+      TestEntity(tString: 'for now') ]);
+
+    await Future.delayed(Duration(seconds: 0)); // ffi explodes without
+    expect(result, [3, 3]);
```
Comment (greenrobot-team) [inline test/stream_test.dart:60]: Same here, shouldn't the counts be `[1, 3]`?

File: test/stream_test.dart (hunk 7 above)
Comment (Buggaboo) [reply test/stream_test.dart:37]: Doh.

File: test/stream_test.dart (hunk 8 above)
Comment (Buggaboo) [reply test/stream_test.dart:60]: Doh again.

File: test/stream_test.dart (hunk 7 above)
Comment (Buggaboo) [reply test/stream_test.dart:37]: I guess for some reason the puts execute completely before the observables fire. Strange behaviour.

File: test/stream_test.dart (hunk 8 above)
Comment (Buggaboo) [reply test/stream_test.dart:60]: I don't know whether this is a dart test async or OB sync issue. I tried setting it up on the flutter demo app, after updating dart/flutter, [but android needs updating again](https://github.com/flutter/flutter/wiki/Upgrading-pre-1.12-Android-projects).

Also there's a non-nullness feature requirement that's blocking me.

Dart SDK version: 2.9.3 (stable) (Tue Sep 8 11:21:00 2020 +0200) on "macos_x64"
Flutter 1.22.0-10.0.pre.384 • channel master • https://github.com/flutter/flutter.git
Framework • revision 062e469f1b (15 hours ago) • 2020-09-27 15:32:07 -0700
Engine • revision 3a73d073c8
Tools • Dart 2.11.0 (build 2.11.0-161.0.dev)

I setup a new flutter android/iOS app and I'll try again.

File: lib/src/observable.dart (hunk 6 above)
Comment (Buggaboo) [reply lib/src/observable.dart]: Fixed. I used `void* user_data` to pass on the store id.

File: lib/src/observable.dart (hunk 1 above)
[the remark this answers is not in the document]
Comment (Buggaboo) [reply lib/src/observable.dart:20]: Hm. Dunno. Copypasta accident maybe.

File: test/stream_test.dart (hunk 8 above)
Comment (Buggaboo) [reply test/stream_test.dart]: You're right; I fixed it with a delay by 0.

File: lib/src/observable.dart [hunk 9]
```diff
@@ -0,0 +1,86 @@
...
+    _anyObserver[storePtr.address] = bindings.obx_observe(storePtr, callback, storePtr);
+  }
+
+  // #53 ffi:Pointer finalizer
+  static void unsubscribe(Store store) {
+    final storeAddress = store.ptr.address;
+    if (!_anyObserver.containsKey(storeAddress)) {
+      return;
+    }
+    bindings.obx_observer_close(_anyObserver[storeAddress]);
```
Comment (greenrobot-team) [inline lib/src/observable.dart:46]: Should wrap in `checkObx()`, similar to `close()` in `store.dart`.

File: test/stream_test.dart [hunk 10]
```diff
@@ -0,0 +1,72 @@
...
+    final condition = text.notNull();
+    final query = box.query(condition).order(text).build();
+    final queryStream = query.findStream();
+    final subscription = queryStream.listen((list) {
+      final str = list.map((t) => t.tString).toList().join(', ');
+      result.add(str);
+    });
+
+    box.put(TestEntity(tString: 'Hello world'));
+    await Future.delayed(Duration(seconds: 0)); // ffi explodes without
```
Comment (greenrobot-team) [inline test/stream_test.dart:32]: The docs for Future.delayed contain

> If the duration is 0 or less, it completes no sooner than in the next event-loop iteration, after all microtasks have run.

I don't know much about Dart, but this sounds like the observer callback is held up from executing while this "thread" is active. Might this cause issues in an actual app where the callback execution is held back too long? Is it possible to run them on a separate "thread" so they are not held up by other code?

Comment (greenrobot-team) [reply test/stream_test.dart:32]: Hm, reading the `StreamController.broadcast` docs on using `sync` is false mode this actually sounds like expected behavior (events may be delivered at any point after the code adding the event has completed). LGTM then.

Comment (greenrobot-team) [reply test/stream_test.dart:32]: Maybe remove that `ffi explodes without` comment than and clarify regarding delayed event delivery.

File: example/flutter/objectbox_demo/lib/main.dart [hunk 11]
```diff
@@ -43,143 +49,169 @@ class MyHomePage extends StatefulWidget {
...
+
+  ViewModel(Directory dir) {
+    _store = Store(getObjectBoxModel(), directory: dir.path + '/objectbox');
+    _box = Box<Note>(_store);
+
+    final dateProp = Note_.date;
+    final dummyCondition = dateProp.greaterThan(0);
+
+    _query = _box
+        .query(dummyCondition)
```
Comment (greenrobot-team) [inline example/flutter/objectbox_demo/lib/main.dart:65]: It's not possible to query without conditions?! We should create an issue for that.

File: example/flutter/objectbox_demo/lib/main.dart [hunk 12]
```diff
@@ -43,143 +49,169 @@ class MyHomePage extends StatefulWidget {
...
+
+  void removeNote(Note note) => _box.remove(note.id);
+
+  get queryStream => _query.findStream();
+
+  get allNotes => _query.find();
+
+  void dispose() {
+    _query.close();
+    _store.unsubscribe();
```
Comment (greenrobot-team) [inline example/flutter/objectbox_demo/lib/main.dart:80]: I wonder if this should be part of `close()`. This feels like something very easy to forget.

**Comment by greenrobot-team** [comment]
@Buggaboo Thanks for your latest changes and 👍 for also improving our example app! I had some more questions, but this LGTM now.

@vaind Can you look at `observable.dart` and check if using `store.prt` as `user_data` when creating an observer is fine? My knowledge is lacking there. (Note: is on holiday, so this might take a while.)

**Comment by greenrobot-team** [comment]
Sorry, forgot one thing: look at Dart Analysis issues and run the formatter.

File: lib/src/observable.dart (hunk 9 above)
Comment (Buggaboo) [reply lib/src/observable.dart:46]: I checked against [this](https://github.com/objectbox/objectbox-c/blob/master/include/objectbox.h#L1479).
The return type is `void`. Should it return anything to the `checkObx`?

I don't know if `latestNativeError` should return anything, if used with this.

File: test/stream_test.dart (hunk 10 above)
Comment (Buggaboo) [reply test/stream_test.dart:32]: I updated this.

File: example/flutter/objectbox_demo/lib/main.dart (hunk 11 above)
Comment (Buggaboo) [reply example/flutter/objectbox_demo/lib/main.dart:65]: Agreed.

File: example/flutter/objectbox_demo/lib/main.dart (hunk 12 above)
Comment (Buggaboo) [reply example/flutter/objectbox_demo/lib/main.dart:80]: It's called by `close` iirc. I didn't want to expose the BLoC/VM's internals to the Widgets.

File: lib/src/observable.dart (hunk 9 above)
Comment (greenrobot-team) [reply lib/src/observable.dart:46]: Ah, sorry. I looked at the internal version of `objectbox.h` which is not yet released. All good!

File: example/flutter/objectbox_demo/lib/main.dart (hunk 12 above)
Comment (greenrobot-team) [reply example/flutter/objectbox_demo/lib/main.dart:80]: Oh, sorry, should have been more clear: was talking about not even exposing `unsubscribe()` to the user and doing it as part of `close()`.

Comment (Buggaboo) [reply example/flutter/objectbox_demo/lib/main.dart:80]: Good point.

File: example/flutter/objectbox_demo/lib/main.dart (hunk 11 above)
Comment (greenrobot-team) [reply example/flutter/objectbox_demo/lib/main.dart:65]: https://github.com/objectbox/objectbox-dart/issues/125

**Comment by greenrobot-team** [comment]
Squashed and merged. I fixed the analysis and formatting issues myself (https://github.com/objectbox/objectbox-dart/commit/0a941a5280a080b45e29c12399033c4912f01b57).

Thanks again.

**Comment by vaind** [comment]
> @vaind Can you look at `observable.dart` and check if using `store.prt` as `user_data` when creating an observer is fine? My knowledge is lacking there. (Note: is on holiday, so this might take a while.)

Should be fine to the best of my knowledge.

**Pull request merged**